### PR TITLE
Create tracking microservice for Artemis

### PR DESCRIPTION
**Pull request by JonasPetry**
This component enables Athene to track result elements during the assessment-process in Artemis.

Test:
1. Spin up the containiers using `docker-compose up --build tracking database
`
2. Spin up Artemis with the text-clustering component
3. Assess a text exercise and get the x-athene-tracking-authorization  header from your network tab
4. Use postman to send a request to this component (http://localhost:8003//tracking/text-exercise-assessment)
4.1 Set the header "x-athene-tracking-authorization" & JWT token
4.2 Set the body , ***Adjust the result id***
```json
{
  "textBlocks": [
    {
      "id": "4f9ef751f45ee89d5aaaccb5db401dda6048b9de",
      "text": "Test",
      "startIndex": 12,
      "endIndex": 16,
      "type": "AUTOMATIC"
    },
    {
      "id": "8c8d2463ec548efca05e66423bee537b6357e880",
      "text": "Test",
      "startIndex": 0,
      "endIndex": 4,
      "type": "AUTOMATIC"
    },
    {
      "id": "960cffe4f5a46dbb3800ef69b480a3d98dad2126",
      "text": "Test",
      "startIndex": 6,
      "endIndex": 10,
      "type": "AUTOMATIC"
    }
  ],
  "participation": {
    "type": "student",
    "id": 1,
    "initializationState": "FINISHED",
    "initializationDate": "2020-07-07T14:34:18.067248+02:00",
    "exercise": {
      "type": "text",
      "id": 1,
      "title": "test",
      "maxScore": 1,
      "mode": "INDIVIDUAL",
      "presentationScoreEnabled": false,
      "course": {
        "id": 1,
        "title": "Test",
        "shortName": "test",
        "studentGroupName": "artemis-dev",
        "teachingAssistantGroupName": "artemis-dev",
        "instructorGroupName": "artemis-dev",
        "onlineCourse": false,
        "maxComplaints": 3,
        "maxTeamComplaints": 3,
        "maxComplaintTimeDays": 7,
        "studentQuestionsEnabled": true,
        "registrationEnabled": false,
        "presentationScore": 0,
        "complaintsEnabled": true
      },
      "automaticAssessmentEnabled": false,
      "teamMode": false,
      "visibleToStudents": true,
      "released": true,
      "studentAssignedTeamIdComputed": false,
      "ended": false
    },
    "results": [
      {
        "id": 2,
        "rated": true,
        "hasFeedback": false,
        "feedbacks": [
          {
            "id": 4,
            "detailText": "test",
            "reference": "8c8d2463ec548efca05e66423bee537b6357e880",
            "credits": 0.5,
            "positive": true,
            "type": "MANUAL"
          }
        ],
        "assessor": {
          "id": 2,
          "login": "ga27der",
          "firstName": "Jonas Petry",
          "email": "jonas.petry@manageandmore.de",
          "activated": false,
          "langKey": "en",
          "name": "Jonas Petry",
          "participantIdentifier": "ga27der"
        },
        "assessmentType": "MANUAL",
        "submission": null,
        "participation": null
      }
    ],
    "submissions": [],
    "student": {
      "id": 2,
      "login": "ga27der",
      "firstName": "Jonas Petry",
      "email": "jonas.petry@manageandmore.de",
      "activated": false,
      "langKey": "en",
      "name": "Jonas Petry",
      "participantIdentifier": "ga27der"
    },
    "participantIdentifier": "ga27der",
    "participantName": "Jonas Petry",
    "trackingToken": "eyJhbGciOiJIUzI1NiJ9.eyJyZXN1bHRfaWQiOjMsImV4cCI6MTU5NDE0OTI0NX0.X43Tmy5NdV07JLZ987Tl-rt6jW7qkPpmt3Nd9xsOrcY"
  }
}
```

See that you get a response

### CODE REVIEW

File: .github/workflows/dockerimage.yml [hunk 1]
```diff
@@ -43,3 +43,15 @@ jobs:
...
+  athene-tracking:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v1
+    - uses: azure/docker-login@v1
+      with:
+        login-server: docker.pkg.github.com
+        username: ${GITHUB_ACTOR}
+        password: ${{ secrets.GITHUB_TOKEN }}
+    - name: Build and Push the artemis-tracking Docker image
```
Comment (jpbernius) [inline .github/workflows/dockerimage.yml:56]:
```suggestion
    - uses: azure/docker-login@v1
      with:
        username: ${{ secrets.DOCKERHUB_USER }}
        password: ${{ secrets.DOCKERHUB_TOKEN }}
      if: github.ref == 'refs/heads/master'
    - name: Build and Push the artemis-tracking Docker image
```

File: tracking/src/main.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+
+@app.post('/tracking/text-exercise-assessment', status_code=201)
+async def testPost(request: Request, response: Response):
+    feedback = await request.json()
+    jwt_token = request.headers.get('x-athene-tracking-authorization')
+    secret_base64 = environ['JWT_SECRET_BASE64']
+    try:
+        encoded_jwt_token = decode(jwt_token, base64.b64decode(secret_base64), verify=True, algorithms=['HS256'])
+        if (encoded_jwt_token.get('result_id')) != feedback.get('participation').get('results')[0].get('id'):
```
Comment (jpbernius) [inline tracking/src/main.py:23]:
```suggestion
        if encoded_jwt_token.get('result_id') != feedback.get('participation').get('results')[0].get('id'):
```

File: tracking/src/main.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+from os import environ
+
+app = FastAPI()
+
+
+@app.get('/tracking')
+async def root():
+    return {'message': 'Hello World!'}
+
+
```
Comment (jpbernius) [inline tracking/src/main.py:15]:
```suggestion
```

**Review by jpbernius — DISMISSED** [review]

File: init-mongo.js [hunk 4]
```diff
@@ -22,3 +22,16 @@ db.createUser(
...
+        user: "tracking",
+        pwd: "tracking_password",
+        roles: [
+            {
+                role: "readWrite",
+                db: "athene_db"
+            }
+        ]
+    }
+);
```
Comment (linusmichel) [inline init-mongo.js:37]: Newline at EOF missing

File: docker-compose.yml [hunk 5]
```diff
@@ -39,6 +39,27 @@ services:
...
+      - DATABASE_PORT=27017
+      - DATABASE_NAME=athene_db
+      - DATABASE_USER=tracking
+      - DATABASE_PWD=tracking_password
+    working_dir: /usr/src/app
+    depends_on:
+      - database
+    networks:
+      - athene
+
```
Comment (linusmichel) [inline docker-compose.yml:62]: We need to add this to the traefik config once the load-balancer branch is merged as well

File: tracking/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,9 @@
+FROM tiangolo/uvicorn-gunicorn-fastapi:python3.7
+
+COPY requirements.txt requirements.txt
+RUN pip install -r requirements.txt
+
+COPY src/ src/
+
+EXPOSE 80
+CMD uvicorn --port 80 --host 0.0.0.0 src.main:app
```
Comment (linusmichel) [inline tracking/Dockerfile:9]: Newline at EOF missing

**Review by linusmichel — CHANGES_REQUESTED** [review]
Looks good to me and works in my environment

**Review by linusmichel — APPROVED** [review]

**Pull request merged**